exercises/grep/grep_test.py

```diff
@@ -36,17 +37,55 @@
 Restore us, and regain the blissful Seat,
 Sing Heav'nly Muse, that on the secret top
 Of Oreb, or of Sinai, didst inspire
-That Shepherd, who first taught the chosen Seed
-'''
+That Shepherd, who first taught the chosen Seed'''
+FILENAMES = [
+    ILIADFILENAME,
+    MIDSUMMERNIGHTFILENAME,
+    PARADISELOSTFILENAME,
+]
+builtins.FILES = {}
 
 
-def remove_file(file_name):
-    try:
-        os.remove(file_name)
-    except OSError:
+class File(object):
+    def __init__(self, name=''):
+        self.name = name
+        self.contents = ''
+
+    def read(self):
+        return self.contents
+
+    def readlines(self):
+        return [l + '\n' for l in self.read().split('\n') if l]
+
+    def write(self, data):
+        self.contents += data
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
         pass
 
 
+builtins.oldopen = builtins.open
+
+
+def open(name, mode='r', *args, **kwargs):
+    if name in FILENAMES:
+        if mode == 'w' or name not in builtins.FILES:
+            builtins.FILES[name] = File(name)
+        return builtins.FILES[name]
+    else:
+        return builtins.oldopen(name, mode, *args, **kwargs)
+
+
+builtins.open = open
+
+
+def remove_file(file_name):
+    del builtins.FILES[file_name]
+
+
```

Comment:
I think it would be useful for these lines (70-88) to have comments explaining what it's doing at each step. It'll just make it easier to deal with if something breaks in future.